### PR TITLE
[CI] Update stats tests

### PR DESCRIPTION
**Pull request by MSP-Greg**
### Description

A few of the stats tests have very long strings/regex's used with assert statements.  These make it difficult to update the tests, and also are quite a pita to work with when a test fails.

The docs state that "[Puma.stats](https://msp-greg.github.io/puma/Puma.html#stats-class_method) produces a JSON string, and the same is returned via `Puma::ControlCLI`.  JSON strings are just that, so we can change the text representation as long as they parse with the same key/value pairs.

Updates the tests to work with 'expectation' hashes, rather than 'expectation' strings/regex's.  Hopefully, others will feel that the updated tests are easier to follow.

Adds a helper method (`assert_hash`) to `test/helpers/test_puma/assertions.rb`

### Your checklist for this pull request
- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added (or updated) appropriate tests if this PR fixes a bug or adds a feature.
- [ ] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [ ] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [ ] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed, including Rubocop.

### CODE REVIEW

**Comment by MSP-Greg** [comment]
@jjb

Since you've been in the stats test code recently, wondering what your thoughts are.

File: test/test_cli.rb [hunk 1]
```diff
@@ -46,6 +46,49 @@ def teardown
     @ready.close
   end
 
+  def check_single_stats(body, check_puma_stats = true)
+    http_hash = JSON.parse body.split("\n", 2).last
```
Comment (jjb) [inline test/test_cli.rb:50]: i might do this in two lines with variable names that indicate what the split and last achieve

File: test/test_integration_pumactl.rb [hunk 2]
```diff
@@ -197,27 +197,80 @@ def test_clustered_stats
     skip_unless :fork
     skip_unless :unix
 
-    puma_version_pattern = "\\d+.\\d+.\\d+(\\.[a-z\\d]+)?"
+    min_threads = 1
+    max_threads = 2
```
Comment (jjb) [inline test/test_integration_pumactl.rb:201]: is this (intentionally?) changing the behavior of the test?

maybe do that in a different standalone PR so the behavior vs. style change is clear

**Review by jjb — COMMENTED** [review]
very nice! couple notes.

> JSON strings are just that, so we can change the text representation as long as they parse with the same key/value pairs

does this refer to the changes in this PR or to future possible changes in app code? the ordering in the app code is hardcoded in worker.rb + worker_handle.rb, and it's parsed without the json library in order to avoid making it a dependency. probably already on your radar but mentioning in case relevant.

File: test/test_integration_pumactl.rb (hunk 2 above)
Comment (MSP-Greg) [reply test/test_integration_pumactl.rb:201]: Yes, current code had the following:
```ruby
'running' => 2,
'pool_capacity'  => 2,
'max_threads'    => 2,
```

PR code has:
```ruby
'running' => 1,
'pool_capacity'  => 2,
'max_threads'    => 2,
```

So, 'running' no longer has the same value as 'max_threads'.  Changing 'pool_capacity' is a bit messier in CI...

File: test/test_cli.rb (hunk 1 above)
Comment (MSP-Greg) [reply test/test_cli.rb:50]: Actually, that isn't needed.  My mistake.  Removed.

The tests in `test_integration_pumactl.rb` use `Puma::ControlCLI` which outputs the JSON string to `@stdout`, but prefaces it with a line "Command stats sent success".

`test_cli.rb` sends a simple request, and only the JSON string is returned.  I was thinking it also had the line that needed to be stripped...

**Comment by MSP-Greg** [comment]
> does this refer to the changes in this PR or to future possible changes in app code?

No changes to lib files in this PR, only test files.  So, the JSON strings have not changed.

My point was that it's a JSON string (IIRC, JSON allows white-space characters in various places), but our tests were treating it as a 'standard' string.  One of the expectation regex strings was almost 370 characters.

By parsing the JSON string, the tests can compare hashes, and are more readable and maintainable.

I pushed another commit fixing the above.

File: test/test_integration_pumactl.rb (hunk 2 above)
Comment (jjb) [reply test/test_integration_pumactl.rb:201]: gotcha - why the change?

File: test/test_cli.rb (hunk 1 above)
Comment (jjb) [reply test/test_cli.rb]: nice

is it valuable to also test the value of `first` as well?

File: test/test_integration_pumactl.rb (hunk 2 above)
Comment (MSP-Greg) [reply test/test_integration_pumactl.rb:201]: > why the change?

So fewer of the hash values have the same value.  Current code has three with the same value of 2. After the PR, only two values are identical.  Really, all hash values need to be unique for the tests to verify that they're all correct.

Off topic, but with some planned test suite updates, we should be able to have a test file something similar to `test_stats`, where all the 'full' stats tests would be located.

File: test/test_cli.rb (hunk 1 above)
Comment (MSP-Greg) [reply test/test_cli.rb]: There's a few lines in `test_integration_pumactl.rb` that check it. The code in `puma/control_cli.rb` is:

https://github.com/puma/puma/blob/36c7307607432416d9a58eef317757da1eceb95a/lib/puma/control_cli.rb#L225-L226

**Comment by MSP-Greg** [comment]
@jjb

> the ordering in the app code is hardcoded in worker.rb + worker_handle.rb, and it's parsed without the json library in order to avoid making it a dependency. probably already on your radar but mentioning in case relevant.

Just noticed the above.  Just to clarify, JSON is loaded in the test process, and has been for a few years.  There is a test that spawns (`IO.popen`) a Puma process and checks that JSON is not loaded.  The sub-process is needed to isolate the Puma instance from the test environment.

https://github.com/puma/puma/blob/36c7307607432416d9a58eef317757da1eceb95a/test/test_integration_cluster.rb#L386-L390

Also, the code in `Puma::JSONSerialization` only contains a public `generate` method, there is no JSON parsing in the Puma lib files.

**Comment by MSP-Greg** [comment]
Just rebased
